components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/OutboundProvisioningManager.java

```diff
@@ -865,16 +876,32 @@ protected boolean canUserBeProvisioned(ProvisioningEntity provisioningEntity,
             // we apply restrictions only for users.
             // if service provider's out-bound provisioning configuration does not define any roles
             // to be provisioned then we apply no restrictions.
+            /*  After group role separation, the internal roles should be not be provisioned. Only groups should be
+                provisioned.
+            */
+            if (isInternalRole(provisioningEntity.getEntityName(), tenantDomain)) {
+                return false;
+            }
             return true;
         }
 
         if (provisioningEntity.getAttributes() != null &&
                 StringUtils.isNotBlank(provisioningEntity.getEntityName())) {
             String userName = provisioningEntity.getEntityName();
+            isInternalRole(provisioningEntity.getEntityName(), tenantDomain);
             List<String> provisioningRoleList = Arrays.asList(provisionByRoleList);
+            if (provisioningEntity.getInboundAttributes() == null ||
+                    provisioningEntity.getInboundAttributes().get(USER_ID_CLAIM) == null) {
+                return false;
+            }
+            List<String> roleListOfUser = getUserRoles(provisioningEntity.getInboundAttributes().get(USER_ID_CLAIM),
+                    tenantDomain);
+            List<String> groupRoleListOfUser = new ArrayList<>(roleListOfUser);
 
-            List<String> roleListOfUser = getUserRoles(userName, tenantDomain);
-            if (userHasProvisioningRoles(roleListOfUser, provisioningRoleList, userName)) {
+            List<String> groupListOfUser = getUserGroups(userName, tenantDomain);
+            groupRoleListOfUser.addAll(groupListOfUser);
```

Comment:
Provisioning allowed list contains list of role-uuids and group names
Ex: "2642fb7a-58df-4608-aff9-b90aa73bcc51,group2"

In order to check a matching role or group, the user roles and groups are combined.